numpy/core/src/umath/ufunc_object.c

```diff
@@ -438,8 +471,15 @@ _parse_signature(PyUFuncObject *ufunc, const char *signature)
         while (signature[i] != ')') {
             /* loop over core dimensions */
             int j = 0;
-            if (!_is_alpha_underscore(signature[i])) {
-                parse_error = "expect dimension name";
+            npy_intp frozen_size = -1;
+            if (signature[i] == '\0') {
+                parse_error = "unexpected end of signature string";
+                goto fail;
+            }
+
+            if (!_is_alpha_underscore(signature[i]) &&
```

Comment:
I think here and below it might make more sense to separate out more:
```
if (_is_alpha...) {
    check on var_names 
}
else {
    frozen_size = _get_...
    if (frozen_size < 0) {
        parse error
    }
    check on same frozen size
}
```

Reply:
fixed by removing frozen size code